### PR TITLE
Update clang-format version

### LINKED ISSUE #6381: Update `clang-format` version
Follow-up to #6376, see https://github.com/kokkos/kokkos/issues/6376#issuecomment-1690544398.

It seems that he have agreed on using `constexpr West` in Kokkos. Update `clang-format` version used in CI and adjust `.clang-format` accordingly.

Relevant option: https://clang.llvm.org/docs/ClangFormatStyleOptions.html#qualifierorder.

### PR DESCRIPTION
**Pull request by masterleinad**
Fixes #6381. As discussed in https://kokkosteam.slack.com/archives/G5CBLMFLP/p1721752120238319, using clang-format 8 has become pretty annoying since that particular version is pretty hard to obtain. Looking through various package managers it seems that 14, 15, or 16 are good choices. 16 has the problem that it won't be available in Ubuntu 22.04 natively but that's what Kokkos Kernels has chosen in https://github.com/kokkos/kokkos-kernels/pull/2255.
This pull request has commits with all 3 versions so we can also decide if either of the updates in clang-format (14->15, 15->16) would be worth it.

### CODE REVIEW

**Comment by masterleinad** [comment]
14->15:
- Important change: Renamed IndentRequires to IndentRequiresClause and changed the default for all styles from false to true.
- Reworked and improved handling of concepts and requires. Added the RequiresClausePosition option as part of that.
- Changed BreakBeforeConceptDeclarations from Boolean to an enum.
- Option InsertBraces has been added to insert optional braces after control statements.

15->16:
- Add RemoveSemicolon option for removing ; after a non-empty function definition.
- Add RequiresExpressionIndentation option for configuring the alignment of requires-expressions. The default value of this option is OuterScope, which differs in behavior from clang-format 15. To match the default behavior of clang-format 15, use the Keyword value.
- Add IntegerLiteralSeparator option for fixing integer literal separators in C++, C#, Java, and JavaScript.
- Add BreakAfterAttributes option for breaking after a group of C++11 attributes before a function declaration/definition name.
- Add InsertNewlineAtEOF option for inserting a newline at EOF if missing.
- Add LineEnding option to deprecate DeriveLineEnding and UseCRLF.

**Comment by cwpearson** [comment]
I maintain a couple clang-format containers here for use in Kokkos Comm and Kokkos Kernels since there was a bit of concern about blindly running a "third party" container on Sandia filesystems. We use them through podman on some of our systems at Sandia.

https://github.com/cwpearson/clang-format

**Comment by nliber** [comment]
Because we change this so rarely, I'm in favor of going to the latest version of clang-format that we possibly can.

In addition, whenever we have to update most of the files (such as changing the copyright notice on major releases), we should also consider bumping up the clang-format version at that time.

**Comment by masterleinad** [comment]
@jgfouca Was there a particular reason for choosing clang-format 16 in https://github.com/kokkos/kokkos-kernels/pull/2255?

**Comment by masterleinad** [comment]
@ndellingwood Is in favor of doing this before the release.

**Comment by jgfouca** [comment]
@masterleinad , we wanted a version new enough to support newer C++ standards but not so new that it wasn't easily available in package managers. Admittedly, I did not do much of a deep dive into the various versions of clang-format. I did have to bump up the Ubuntu version in the github workflow from 20.04 to 24.04 in order for it to be able to get 16. The latest release is 18, so perhaps we should have just gone all the way up to 18.

**Comment by cz4rs** [comment]
Related issue from the last summer: https://github.com/kokkos/kokkos/issues/6381.

**Comment by nmm0** [comment]
Clang-format 16 can be installed on ubuntu 22.04 via https://apt.llvm.org/ (which likely you want anyway if you are planning on doing a lot of clang development)

**Comment by aprokop** [comment]
Is there an appetite to increasing the number of characters per line to something like 99?

**Comment by cedricchevalier19** [comment]
> Is there an appetite to increasing the number of characters per line to something like 99?

Kokkos-Kernels and Kokkos-Comm have set their limits to 120.

**Comment by masterleinad** [comment]
> Kokkos-Kernels and Kokkos-Comm have set their limits to 120.

That would imply
```
 942 files changed, 33777 insertions(+), 62760 deletions(-)
```

**Comment by lucbv** [comment]
I will advocate to have the same one in Core and Kernels so clang-16, others have explained our rational around this choice (new, supports recent c++ standards, reasonably available).
Same goes with picking line length of 120 characters, no particular reason beyond the obvious ease of read and overall line reduction I guess?

**Comment by masterleinad** [comment]
I'm not a big fan of increasing the line length but I also think this could be a separate discussion since the changes in the current form are relatively few.

**Comment by masterleinad** [comment]
clang-15 still seems to have more compatibility with relevant OS but I can see the point in using clang-16 instead.

**Comment by Rombur** [comment]
I am in favor of using Clang 16 since that's what Kokkos Kernels is using.

**Comment by aprokop** [comment]
Is there a guarantee for compatibility among different clang-format versions with the same major version? E.g., are all 16.y.z compatible?

**Comment by masterleinad** [comment]
> Is there a guarantee for compatibility among different clang-format versions with the same major version? E.g., are all 16.y.z compatible?

I would hope so but I would be surprised if there was any difference in that respect between major versions.

**Comment by nliber** [comment]
I'm in favor of increasing the line length.

Will clang-format re-indent when we do that?

**Comment by masterleinad** [comment]
> Will clang-format re-indent if we do that?

Yes, see https://github.com/kokkos/kokkos/pull/7150#issuecomment-2262953118.

**Comment by masterleinad** [comment]
The last commit uses `QualifierAlignment: Left` since that is the style we are mostly following and we need to decide on the position of the `const` qualifier when enforcing "`constexpr` west" (https://github.com/kokkos/kokkos/pull/6376).

**Comment by masterleinad** [comment]
I used clang-format 16.0.6 locally and both GitHub CI and jenkins are fine with the formatting.

File: scripts/docker/Dockerfile.clang [hunk 1]
```diff
@@ -1,13 +1,13 @@
-FROM ubuntu:20.04
+FROM ubuntu:24.04
 
 RUN apt-get update && apt-get install -y \
         bc \
         git \
         build-essential \
-        clang-format-8 \
+        clang-format-16 \
         wget \
```
Comment (aprokop) [inline scripts/docker/Dockerfile.clang:8]: I think you only need to install `clang-format-16` and nothing else.

**Review by dalg24 — APPROVED** [review]
Approving the bumping to clang-format-16 with no change to configuration file

**Comment by aprokop** [comment]
Needs rebasing

**Review by dalg24 — COMMENTED** [review]
You need to drop all changes but the configuration file header and the workflow, rebase onto develop and apply the format.

They are a bunch of conflicts and changes to the QualifierAlignment are still present.

**Review by aprokop — APPROVED** [review]

**Comment by dalg24** [comment]
Failures were at runtime (GPU not detected)

**Pull request merged**